tool/tsh/kubectl.go

```diff
@@ -0,0 +1,272 @@
+	cmd.Env = append(os.Environ(), fmt.Sprintf("%s=yes", tshKubectlReexec))
+	return trace.Wrap(cmd.Run())
+}
+
+// runKubectlCode runs the actual kubectl package code with the default options.
+// This code is only executed when `tshKubectlReexec` env is present. This happens
+// because we need to retry kubectl calls and `kubectl` calls os.Exit in multiple
+// paths.
+func runKubectlCode(args []string) {
+	// These values are the defaults used by kubectl and can be found here:
+	// https://github.com/kubernetes/kubectl/blob/3612c18ed86fc0a2f4467ca355b3e21569fabe0a/pkg/cmd/cmd.go#L94
+	defaultConfigFlags := genericclioptions.NewConfigFlags(true).
+		WithDeprecatedPasswordFlag().
+		WithDiscoveryBurst(300).
+		WithDiscoveryQPS(50.0)
+
+	command := cmd.NewDefaultKubectlCommandWithArgs(
+		cmd.KubectlOptions{
+			// init the default plugin handler.
+			PluginHandler: cmd.NewDefaultPluginHandler(plugin.ValidPluginFilenamePrefixes),
+			Arguments:     args,
+			ConfigFlags:   defaultConfigFlags,
+			// init the IOSStreams.
+			IOStreams: genericclioptions.IOStreams{In: os.Stdin, Out: os.Stdout, ErrOut: os.Stderr},
+		},
+	)
+	// override args without kubectl to avoid errors.
+	command.SetArgs(args[1:])
+	// run command until it finishes.
+	if err := cli.RunNoErrOutput(command); err != nil {
+		// Pretty-print the error and exit with an error.
+		cmdutil.CheckErr(err)
+	}
+	os.Exit(0)
+}
+
+func runKubectlAndCollectRun(cf *CLIConf, args []string) error {
+	var (
+		alreadyRequestedAccess bool
+		err                    error
+		exitErr                *exec.ExitError
+	)
+	for {
+		// missingKubeResources will include the Kubernetes Resources whose access
+		// was rejected in this kubectl call.
+		missingKubeResources := make([]resourceKind, 0, 50)
+		reader, writer := io.Pipe()
+		wg := sync.WaitGroup{}
+		wg.Add(1)
+		go func() {
+			// This goroutine scans each line of output emitted to stderr by kubectl
+			// and parses it in order to check if the returned error was a problem with
+			// missing access level. If it's the case, tsh kubectl will create automatically
+			// the access request for the user to access the resource.
+			// Current supported resources:
+			// - pod
+			// - kube_cluster
+			defer wg.Done()
+
+			scanner := bufio.NewScanner(reader)
+			scanner.Split(bufio.ScanLines)
+			for scanner.Scan() {
+				line := scanner.Text()
+
+				// Check if the request targeting a pod endpoint was denied due to
+				// Teleport Pod RBAC or if the operation was denied by Kubernetes RBAC.
+				// In the second case, we should create a Resource Access Request to allow
+				// the user to exec/read logs using different Kubernetes RBAC principals.
+				// using different Kubernetes RBAC principals.
+				if podForbiddenRe.MatchString(line) {
+					results := podForbiddenRe.FindStringSubmatch(line)
+					missingKubeResources = append(missingKubeResources, resourceKind{kind: types.KindKubePod, subResourceName: filepath.Join(results[2], results[1])})
```

Comment:
Extracting pod name from error message feels a bit fragile but I'm not sure what the other good way to get it would be. Since this is `tsh`, can it maybe extract it from the CLI arguments passed to the command? Cause they should have the pod name already, right? E.g. `tsh kubectl exec -ti nginx`.

Reply:
My idea here was to support cases where users try to edit/combine multiple pod logs at the same time. For those cases, kubectl prints one error message per pod and we just need to collect them and spawn a new process after approval.

When you list pods by selector -i.e. extracting logs from multiple pods at the same time - we don't have access to pod names, we just have access to the field selector or label selector since kubectl does a list request followed by a logs request per pod.

The main use case is when you have list permissions but you don't have the edit/log permissions defined in your Kubernetes Groups and you must create an access request to grant them.

Parsing the cli is also difficult. Although we can parse the cli with kubectl command, we still need to extract that info from the parsed object itself, which means we need to know all the different names they give to pod name args - we still need to get pod names when they are not available.

I think exacting that info from the error message when Teleport and Kubernetes API use the same message format for denied requests is the simplest way to achieve it.